### PR TITLE
Added the TEST to check TFLite sqrt returning NaN for +inf input Issue #120303

### LINKED ISSUE #120303: TFLite sqrt returns NaN for +inf input while Keras returns +inf
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

2.21.0

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Mobile device

_No response_

### Python version

3.11

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

### Current behavior?

TensorFlow Lite produces different results from Keras for tf.sqrt when the input contains positive infinity.

For sqrt(+inf), Keras returns +inf, while the converted TFLite model returns NaN. Finite values in the same tensor are handled correctly.

In the reproducer below, Keras returns [inf, 3, 4, inf], while TFLite returns [nan, 3, 4, nan].

Expected behavior?

The converted TFLite model should match Keras behavior for tf.sqrt on positive infinity inputs. sqrt(+inf) should return +inf, not NaN.

### Standalone code to reproduce the issue

```shell
import os
os.environ["TF_CPP_MIN_LOG_LEVEL"] = "3"

import numpy as np
import tensorflow as tf


class M(tf.keras.Model):
    def call(self, x):
        return tf.sqrt(x)


model = M()
model.build((None, 4))

tflite_model = tf.lite.TFLiteConverter.from_keras_model(model).convert()

x = np.array([[np.inf, 9.0, 16.0, np.inf]], dtype=np.float32)

keras_out = model(x).numpy()

interpreter = tf.lite.Interpreter(model_content=tflite_model)
interpreter.allocate_tensors()
interpreter.set_tensor(interpreter.get_input_details()[0]["index"], x)
interpreter.invoke()

tflite_out = interpreter.get_tensor(interpreter.get_output_details()[0]["index"])

print("input :", x)
print("eager :", keras_out)
print("tflite:", tflite_out)

if np.isposinf(keras_out[0, 0]) and np.isposinf(keras_out[0, 3]) and np.isnan(tflite_out[0, 0]) and np.isnan(tflite_out[0, 3]):
    print("BUG REPRODUCED: TFLite sqrt returns NaN for +inf input")
else:
    print("not reproduced")
```

### Relevant log output

```shell
input : [[inf 9. 16. inf]]
eager : [[inf 3. 4. inf]]
tflite: [[nan 3. 4. nan]]
BUG REPRODUCED: TFLite sqrt returns NaN for +inf input
```

### PR DESCRIPTION
**Pull request by Tanishq216**
Fixes #120303

### What does this PR do?
This PR adds a comprehensive `SqrtEdgeCases` C++ unit test to `elementwise_test.cc` to explicitly enforce IEEE 754 mathematical correctness for the `SQRT` operator across standard edge cases (`-0.0`, `+0.0`, `+inf`, `-inf`, `NaN`).

Currently, TFLite fails this test because `+inf` inputs evaluate to `NaN` instead of `+inf`.

### Important Note for Reviewers regarding XNNPACK
During my investigation, I discovered that the root cause of this bug lies entirely within the `google/XNNPACK` backend delegate. Because XNNPACK compiles with `-ffast-math`, it suffers from two identical bugs:
1. **Scalar (`f32-scalar.h`)**: The fallback `sqrtf()` evaluates to `NaN` due to the compiler assuming non-infinite inputs.
2. **SIMD (`simd-rsqrt.c.in`)**: The high-performance kernels approximate square roots using Newton-Raphson (`x * rsqrt(x)`). When `x = +inf`, the multiplication becomes `+inf * 0.0`, which evaluates to `NaN`.

*Note: I initially added a manual `+inf` branch in `elementwise.cc`, but reverted it because checking for `+inf` element-by-element breaks TFLite loop auto-vectorization and degrades CPU performance. The fix must occur in the delegate.*

**I am submitting a separate upstream PR to `google/XNNPACK` to mask out `+inf` inputs in `simd-rsqrt.c.in`.** 

In the meantime, this TFLite PR adds the missing C++ tests to officially document the expected IEEE 754 behavior. Once my upstream XNNPACK fix is merged and TensorFlow bumps its XNNPACK dependency, this unit test will naturally start passing.

### CODE REVIEW

File: tensorflow/lite/kernels/elementwise.cc [hunk 1]
```diff
@@ -401,7 +401,9 @@ TfLiteStatus SqrtEval(TfLiteContext* context, TfLiteNode* node) {
   const TfLiteType type = input->type;
   switch (type) {
     case kTfLiteFloat32:
-      return EvalNumeric(context, node, std::sqrt);
+      return EvalNumeric(context, node, [](float f) {
+        return f == std::numeric_limits<float>::infinity() ? f : std::sqrt(f);
+      });
```
[the remark this answers is not in the document]
Comment (Tanishq216) [reply tensorflow/lite/kernels/elementwise.cc]: You are entirely correct. After deeper investigation, I discovered the root cause is actually in google/XNNPACK (specifically simd-rsqrt.c.in multiplying +inf * 0.0 which produces NaN). I will revert the change to elementwise.cc here so it doesn't hurt vectorization, but I will leave the new SqrtEdgeCases C++ test in this PR to prevent future regressions. I am preparing a separate PR for the XNNPACK repository to fix the core -ffast-math vectorization bug!

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Hi @Tanishq216,

Thanks for the contribution and the detailed analysis regarding the XNNPACK behavior!

When testing this CL inside Google's internal environment, we noticed a minor issue with the new SqrtEdgeCases test:

In TensorFlow Lite, denormal floating-point numbers are flushed to zero (FTZ) for performance reasons (using ruy::ScopedSuppressDenormals). Because of this, the input std::numeric_limits<float>::denorm_min() gets flushed to 0.0f during kernel execution, which results in std::sqrt(0.0f) = 0.0f and makes the assertion EXPECT_GT(output[7], 0.0f) fail.

Could you please update the test payload to use std::numeric_limits<float>::min() (the smallest normalized float) instead of denorm_min()? It will prevent the value from being flushed to zero and ensure the test passes reliably across different hardware environments and compilation flags.

Also, thank you for tracking down the XNNPACK bug. We would highly welcome a separate PR to google/XNNPACK to fix the NaN return for +inf when you are ready.

Once the unit test is updated and tests are green, we will trigger the internal import pipeline to merge your changes. Thank you!

**Comment by Tanishq216** [comment]
@dmiltr3 Thanks for the review and I have made the necessary changes as requested. I also would like to inform that there is already a separate PR awaiting review in XNNPACK repository.

**Review by dmiltr3 — APPROVED** [review]

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
..
